float32 loc_x
float32 loc_y
float32 loc_z
